### PR TITLE
Flush log messages and other improvements

### LINKED ISSUE #33: [feature request] Ability to flush logs
I was just debugging a startup issue in our program where it locks up. It was hard to figure out from the logs (configured to output to both Console and GELF) where it got to because several messages remain in the buffer and are never flushed out.

It would be nice if I could manually flush the logs out, or even better if the logging thread flushed them after a certain timeout. I would be happy to prepare a PR if you are interested but I would probably need a little guidance around the best way to add it to the current architecture.

### PR DESCRIPTION
**Pull request by SkyToGround**
### Description of work

Things done:
* Added performance tests.
* Replaced the home-brewed concurrent queue with a *much* faster open source one.
* Made the logger run in a separate thread.
* Added flushing functionality, see documentation for details in how it works.
* Added support for the fmtlib (for formatting) if present. *Note:* The formatting is done in a separate thread which should in most cases be faster than doing it in the thread that calls the logging library.
* Improved/fixed the README-file.
* Made it easier/possible to build the library without using Conan for providing dependencies.
* Added and fixed examples.
* Fixed the non-working `emptyQueue()` and `queueSize()` implementations.
* Fixed/improved code documentation.
* De-activated the use of clang-format in our continuous integration system. It will be re-activated when we have updated clang-format on our build nodes to a newer version.
* Removed the message queue size parameter from log message handlers as it was deemed unnecessary.
* Fixed and added unit tests.

This PR makes some minor changes to the interface that are not backwards compatible, hence we change the version number to 2.0.0. For most users of this library however, you should not have to make changes to your code.

### Issue

* Closes #33 

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel.

### CODE REVIEW

File: include/graylog_logger/FileInterface.hpp [hunk 1]
```diff
@@ -22,6 +22,24 @@ class FileInterface : public BaseLogHandler {
...
   void addMessage(const LogMessage &Message) override;
 
+  /// \brief Waits for all messages created before the call to flush to be
+  /// printed and then flushes the file stream.
+  /// \param[in] TimeOut Amount of time to wait for messages to be written.
+  /// \return Returns true if queue was emptied and stream flushed before the
+  /// time out. Returns false otherwise.
+  bool flush(std::chrono::system_clock::duration TimeOut) override;
+
+  /// \brief Are there any more queued messages?
```
Comment (mattclarke) [inline include/graylog_logger/FileInterface.hpp]:
```suggestion
  /// \brief Are there any queued messages?
```

File: include/graylog_logger/GraylogInterface.hpp [hunk 2]
```diff
@@ -30,14 +31,31 @@ class GraylogConnection {
...
+  /// \brief Waits for all messages created before the call to flush to be
+  /// transmitted.
+  /// \param[in] TimeOut Amount of time to wait for messages to be transmitted.
+  /// \return Returns true if messages were transmitted before the time out.
+  /// Returns false otherwise.
+  /// \note It is possible (though unlikely) that not all log messages have
+  /// been transmitted even if flush() returns true.
+  bool flush(std::chrono::system_clock::duration TimeOut) override;
+
+  /// \brief Are there any more queued messages?
```
Comment (mattclarke) [inline include/graylog_logger/GraylogInterface.hpp]:
```suggestion
  /// \brief Are there any queued messages?
```

File: include/graylog_logger/LogUtil.hpp [hunk 3]
```diff
@@ -108,24 +108,32 @@ class BaseLogHandler {
...
   /// \param[in] MaxQueueLength The maximum number of log messages stored.
-  explicit BaseLogHandler(const size_t MaxQueueLength = 100);
+  BaseLogHandler() = default;
   virtual ~BaseLogHandler() = default;
+
   /// \brief Called by the logging library when a new log message is created.
-  /// \note If the queue of messages is full, any new messages are discarded
-  /// without any indication that this has been done.
+  ///
+  /// Must be implement by any derived classes.
```
Comment (mattclarke) [inline include/graylog_logger/LogUtil.hpp]:
```suggestion
```
Enforced by the compiler, so unnecessary?

Comment (SkyToGround) [reply include/graylog_logger/LogUtil.hpp]: Right.
```suggestion
```

File: include/graylog_logger/LoggingBase.hpp [hunk 4]
```diff
@@ -9,39 +9,132 @@
...
+    auto ThreadId = std::this_thread::get_id();
+    auto UsedArguments = std::make_tuple(args...);
+    Executor.SendWork([=]() {
+      if (int(Level) > int(MinSeverity)) {
+        return;
+      }
+      LogMessage cMsg(BaseMsg);
+      cMsg.SeverityLevel = Level;
+      cMsg.Timestamp = std::chrono::system_clock::now();
+      auto format_message = [=, &cMsg](const auto &... args) {
```
Comment (mattclarke) [inline include/graylog_logger/LoggingBase.hpp]: Looks like you only need Format, so perhaps take that by ref too?
```suggestion
      auto format_message = [&Format, &cMsg](const auto &... args) {
```

File: include/graylog_logger/LoggingBase.hpp [hunk 5]
```diff
@@ -9,39 +9,132 @@
...
+    Executor.SendWork([=]() { BaseMsg.addField(Key, Value); });
   };
   virtual void removeAllHandlers();
   virtual void setMinSeverity(Severity Level);
   virtual std::vector<LogHandler_P> getHandlers();
 
+  virtual bool flush(std::chrono::system_clock::duration TimeOut) {
+    auto FlushCompleted = std::make_shared<std::promise<bool>>();
+    auto FlushCompletedValue = FlushCompleted->get_future();
+    Executor.SendWork([ =, FlushCompleted{std::move(FlushCompleted)} ]() {
```
Comment (mattclarke) [inline include/graylog_logger/LoggingBase.hpp]:
```suggestion
    Executor.SendWork([=, FlushCompleted{std::move(FlushCompleted)} ]() {
```

File: src/LoggingBase.cpp [hunk 6]
```diff
@@ -96,79 +94,35 @@ std::string get_process_name() {
...
-    const Severity Level, const std::string &Message,
-    const std::pair<std::string, AdditionalField> &ExtraField) {
-  log(Level, Message, std::vector<std::pair<std::string, AdditionalField>>{
-                          ExtraField,
-                      });
+  Executor.SendWork([=]() {
+    const int StringBufferSize = 100;
+    std::array<char, StringBufferSize> StringBuffer{};
+    int res;
+    res =
```
Comment (mattclarke) [inline src/LoggingBase.cpp]:
```suggestion
    const int res =
```

File: src/LoggingBase.cpp [hunk 7]
```diff
@@ -96,79 +94,35 @@ std::string get_process_name() {
...
-void LoggingBase::log(
-    const Severity Level, const std::string &Message,
-    const std::pair<std::string, AdditionalField> &ExtraField) {
-  log(Level, Message, std::vector<std::pair<std::string, AdditionalField>>{
-                          ExtraField,
-                      });
+  Executor.SendWork([=]() {
+    const int StringBufferSize = 100;
+    std::array<char, StringBufferSize> StringBuffer{};
+    int res;
```
Comment (mattclarke) [inline src/LoggingBase.cpp]:
```suggestion
```

File: src/LoggingBase.cpp [hunk 8]
```diff
@@ -96,79 +94,35 @@ std::string get_process_name() {
...
+    res =
+        gethostname(static_cast<char *>(StringBuffer.data()), StringBufferSize);
+    if (0 == res) {
+      BaseMsg.Host = std::string(static_cast<char *>(StringBuffer.data()));
+    }
+    BaseMsg.ProcessId = getpid();
+    BaseMsg.ProcessName = get_process_name();
+    // We want to wait for this to finnish, make it so
```
Comment (mattclarke) [inline src/LoggingBase.cpp]:
```suggestion
    // We want to wait for this to finish, make it so
```

File: unit_tests/ConsoleInterfaceTest.cpp [hunk 9]
```diff
@@ -42,3 +43,50 @@ TEST(ConsoleInterface, ConsoleStringFunctionTest) {
   std::string output = testing::internal::GetCapturedStdout();
   ASSERT_EQ(std::string("ALERT: Some test string\n"), output);
 }
+
+TEST(ConsoleInterface, QueueSizeEmpty) {
```
Comment (mattclarke) [inline unit_tests/ConsoleInterfaceTest.cpp]:
```suggestion
TEST(ConsoleInterface, OnInitialisationQueueEmpty) {
```

File: unit_tests/ConsoleInterfaceTest.cpp [hunk 10]
```diff
@@ -42,3 +43,50 @@ TEST(ConsoleInterface, ConsoleStringFunctionTest) {
...
+  ASSERT_EQ(cInter.queueSize(), 0);
+  ASSERT_TRUE(cInter.emptyQueue());
+}
+
+class ConsoleInterfaceStandIn : public ConsoleInterface {
+public:
+  using ConsoleInterface::Executor;
+};
+
+TEST(ConsoleInterface, QueueSizeOne) {
```
Comment (mattclarke) [inline unit_tests/ConsoleInterfaceTest.cpp]:
```suggestion
TEST(ConsoleInterface, QueueSizeOneIsNotEmpty) {
```

File: unit_tests/FileInterfaceTest.cpp [hunk 11]
```diff
@@ -106,3 +107,45 @@ TEST_F(FileInterfaceTest, OpenFileMessages) {
   EXPECT_NE(StdOutputString.find("Unable to open log file"), std::string::npos)
       << "Actual std-output was: " << StdOutputString;
 }
+
+TEST_F(FileInterfaceTest, QueueSizeEmpty) {
```
Comment (mattclarke) [inline unit_tests/FileInterfaceTest.cpp]:
```suggestion
TEST_F(FileInterfaceTest, OnInitialisationQueueEmpty) {
```

File: unit_tests/FileInterfaceTest.cpp [hunk 12]
```diff
@@ -106,3 +107,45 @@ TEST_F(FileInterfaceTest, OpenFileMessages) {
...
+  FileInterfaceStandIn cInter(usedFileName);
+  ASSERT_EQ(cInter.queueSize(), 0);
+  ASSERT_TRUE(cInter.emptyQueue());
+}
+
+TEST_F(FileInterfaceTest, QueueSizeOne) {
```
Comment (mattclarke) [inline unit_tests/FileInterfaceTest.cpp]:
```suggestion
TEST_F(FileInterfaceTest, QueueSizeOneIsNotEmpty) {
```

File: unit_tests/GraylogInterfaceTest.cpp [hunk 13]
```diff
@@ -328,7 +330,36 @@ TEST_F(GraylogConnectionCom, DISABLED_MultipleCloseConnectionTest) {
...
   }
 }
+
+TEST(GraylogInterface, QueueSizeEmpty) {
+  GraylogInterfaceStandIn cInter("localhost", testPort, 10);
+  ASSERT_EQ(cInter.queueSize(), 0);
+  ASSERT_TRUE(cInter.emptyQueue());
+}
+
+TEST(GraylogInterface, QueueSizeOne) {
```
Comment (mattclarke) [inline unit_tests/GraylogInterfaceTest.cpp]:
```suggestion
TEST(GraylogInterface, QueueSizeOneIsNotEmpty) {
```

File: unit_tests/GraylogInterfaceTest.cpp [hunk 14]
```diff
@@ -328,7 +330,36 @@ TEST_F(GraylogConnectionCom, DISABLED_MultipleCloseConnectionTest) {
       logServer->CloseAllConnections();
       std::this_thread::sleep_for(20ms);
     }
-    std::this_thread::sleep_for(1000ms);
+    con.flush(10000ms);
     EXPECT_EQ(logServer->GetNrOfMessages(), NrOfMessages);
   }
 }
+
+TEST(GraylogInterface, QueueSizeEmpty) {
```
Comment (mattclarke) [inline unit_tests/GraylogInterfaceTest.cpp]:
```suggestion
TEST(GraylogInterface, OnInitialisationQueueEmpty) {
```

File: include/graylog_logger/LoggingBase.hpp [hunk 15]
```diff
@@ -9,39 +9,132 @@
...
+        });
+  }
+
+#ifdef WITH_FMT
+  template <typename... Args>
+  void fmt_log(const Severity Level, std::string Format, Args... args) {
+    auto ThreadId = std::this_thread::get_id();
+    auto UsedArguments = std::make_tuple(args...);
+    Executor.SendWork([=]() {
+      if (int(Level) > int(MinSeverity)) {
```
[the remark this answers is not in the document]
Comment (SkyToGround) [reply include/graylog_logger/LoggingBase.hpp]: Maybe, though I am not sure as integers are very cheap in comparison everything else that is going on here. The reason for me doing it this way was to preserve the order in which things are processed. Modification of `MinSeverity` is done in the logging thread which means that if you change severity and then create a log message, the severity might not be updated to properly pass/fail that particular message.

File: include/graylog_logger/LoggingBase.hpp (hunk 4 above)
Comment (SkyToGround) [reply include/graylog_logger/LoggingBase.hpp]: Right!

File: include/graylog_logger/LoggingBase.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (SkyToGround) [reply include/graylog_logger/LoggingBase.hpp]: I thought of it the wrong way. Let me create some benchmarks for testing the performance when the messages have different severity levels. I should be done with that tomorrow.

[the remark this answers is not in the document]
Comment (SkyToGround) [reply include/graylog_logger/LoggingBase.hpp]: When benchmarking, I see no difference in performance for messages that will be accepted. If there are messages that are not accepted however (log level is set to "info" or lower), then there is a performance gain to checking the level before sending off the log message.

I have made made the relevant changes to the code.

**Pull request merged**